fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/ParameterHashUtil.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.persistence.jdbc.util;
+
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Base64.Encoder;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+import com.ibm.fhir.persistence.jdbc.dto.ExtractedParameterValue;
+
+/**
+ * Utility methods for generating Base64-encoded SHA-256 hash for search parameters.
+ */
+public class ParameterHashUtil {
+    private static final Charset UTF_8 = StandardCharsets.UTF_8;
+    private static final String SHA_256 = "SHA-256";
+    private final Encoder encoder;
+    private final MessageDigest digest;
+
+    public ParameterHashUtil() {
+        encoder = Base64.getEncoder();
+        try {
+            digest = MessageDigest.getInstance(SHA_256);
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("MessageDigest not found: " + SHA_256, e);
+        }
+    }
+
+    /**
+     * Gets a Base64-encoded SHA-256 hash for a list of ExtractedParameterValues.
+     * This is used to quickly determine if the search parameters changed
+     * from the existing values in the database, which can then be used to
+     * avoid deleting and re-inserting the search parameters into the database.
+     * @param parameters extracted search parameters
+     * @return the Base64-encoded SHA-256 hash
+     */
+    public String getParametersHash(List<ExtractedParameterValue> parameterValues) {
+        // Sort hashes to make it deterministic
+        List<String> sortedList = new ArrayList<>(parameterValues.size());
+        for (ExtractedParameterValue parameterValue : parameterValues) {
+            sortedList.add(Objects.toString(parameterValue.getHash(this), ""));
+        }
+        sortedList.sort(Comparator.comparing(String::toString));
```

Comment:
I think this algorithm is incorrect. This is sorting on the values of each parameter. The list should be sorted on the parameter names first, and then compute the hash. The current implementation may compute the same hash for parameter sets which are different.

Reply:
The hash is computed from the name/version/url + value, so I thought that would be ok.

Reply:
Updated to generate the hash faster by using a visitor.